### PR TITLE
Added deck_uri to NodeExecutionEvent

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Added deck metadata in TaskMetadata.
https://docs.google.com/document/d/1hIzKEZuk3EQux0R5OJDAYUQmqMF1bbNz7-fAHl1vwH0/edit

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2175

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/core/tasks.proto [hunk 1]
```diff
@@ -95,6 +103,9 @@ message TaskMetadata {
 
     // Indicates whether the system should attempt to execute discoverable instances in serial to avoid duplicate work
     bool cache_serializable = 9;
+
+    // Extra metadata about the deck
+    DeckMetadata deck = 10;
```
Comment (kumare3) [inline protos/flyteidl/core/tasks.proto:108]: we cannot add this here. It is completely dynamic for every task execution. It should be like an output

**Comment by kumare3** [comment]
@pingsutw can you write a one pager on the design and we can then implement it?

File: protos/flyteidl/admin/node_execution.proto [hunk 2]
```diff
@@ -211,6 +211,10 @@ message NodeExecutionGetDataResponse {
     // Full_outputs will only be populated if they are under a configured size threshold. 
     core.LiteralMap full_outputs               = 4;
 
+    // String location uniquely identifying where the deck HTML file is
+    // Should start with the storage location (e.g. s3://, gs:// etc.)
+    string deck_uri                            = 5;
```
Comment (EngHabu) [inline protos/flyteidl/admin/node_execution.proto:216]: How will this be used? is there a design doc (1-pager) I missed? should this be a pre-signed url instead? 
If we have a design doc, do you mind linking it in the PR? if not, do you mind writing a quick 1-pager of how this all fits together?

Comment (pingsutw) [reply protos/flyteidl/admin/node_execution.proto:216]: It's a pre-signed url here. 
https://docs.google.com/document/d/1hIzKEZuk3EQux0R5OJDAYUQmqMF1bbNz7-fAHl1vwH0/edit

TLDR, flyteconsole will retrieve `deck_uri` (pr-signed URL) from `NodeExecutionGetDataResponse`, and then use ifrmae to display the deck.
Something like, 
```
<iframe src="s3://bucket/deck.html">
```

Comment (pingsutw) [reply protos/flyteidl/admin/node_execution.proto:216]: @EngHabu mind taking a look

**Review by EngHabu — DISMISSED** [review]

**Comment by pingsutw** [comment]
cc @wild-endeavor @kumare3 could I get +1 for this?

File: protos/flyteidl/admin/node_execution.proto [hunk 3]
```diff
@@ -211,6 +215,10 @@ message NodeExecutionGetDataResponse {
     // Full_outputs will only be populated if they are under a configured size threshold. 
     core.LiteralMap full_outputs               = 4;
 
+    // String location uniquely identifying where the deck HTML file is.
+    // NativeUrl specifies the url in the format of the configured storage provider (e.g. s3://my-bucket/randomstring/suffix.tar)
+    string deck_uri                            = 5;
```
Comment (EngHabu) [inline protos/flyteidl/admin/node_execution.proto:220]: Why do we need the URI in both Closure and GetDataResponse? isn't it only needed in one?

File: protos/flyteidl/service/dataproxy.proto [hunk 4]
```diff
@@ -44,6 +44,25 @@ message CreateUploadLocationRequest {
...
+  // +optional. The default value comes from a global config.
+  google.protobuf.Duration expires_in = 2;
+
+}
+
+message CreateDownloadLocationResponse {
+  // SignedUrl specifies the url to use to upload content to (e.g. https://my-bucket.s3.amazonaws.com/randomstring/suffix.tar?X-...)
+  string signed_url = 1;
+  // ExpiresAt defines when will the signed URL expires.
+  google.protobuf.Timestamp expires_at = 3;
```
Comment (EngHabu) [inline protos/flyteidl/service/dataproxy.proto:63]:
```suggestion
  google.protobuf.Timestamp expires_at = 2;
```

File: protos/flyteidl/admin/node_execution.proto (hunk 3 above)
Comment (pingsutw) [reply protos/flyteidl/admin/node_execution.proto:220]: `deck_uri` is saved in the Node Execution table and in the closure column, so we have to add it in closure to unmarshal it. https://github.com/pingsutw/flyteadmin/blob/a3f3e4e4e626233f3e79596679cf047d0a1e13a9/pkg/repositories/transformers/node_execution.go#L271

File: protos/flyteidl/service/dataproxy.proto (hunk 4 above)
Comment (pingsutw) [reply protos/flyteidl/service/dataproxy.proto]: sorry, my bad, fixed it.

File: protos/flyteidl/admin/node_execution.proto (hunk 3 above)
Comment (EngHabu) [reply protos/flyteidl/admin/node_execution.proto:220]: aha... ic... then I think we shouldn't have it here... we should just return it in the node execution closure... this is not "data" anyway...

Comment (pingsutw) [reply protos/flyteidl/admin/node_execution.proto]: ic, removed it.

**Review by EngHabu — DISMISSED** [review]

File: protos/flyteidl/service/dataproxy.proto [hunk 5]
```diff
@@ -44,6 +44,25 @@ message CreateUploadLocationRequest {
...
+  string native_url = 1;
+
+  // ExpiresIn defines a requested expiration duration for the generated url. The request will be rejected if this
+  // exceeds the platform allowed max.
+  // +optional. The default value comes from a global config.
+  google.protobuf.Duration expires_in = 2;
+
+}
+
+message CreateDownloadLocationResponse {
```
Comment (kumare3) [inline protos/flyteidl/service/dataproxy.proto:59]: Should this also not say permission denied or is that an error

Comment (kumare3) [reply protos/flyteidl/service/dataproxy.proto:59]: If an error should it not be explicitly called

File: protos/flyteidl/event/event.proto [hunk 6]
```diff
@@ -95,6 +95,10 @@ message NodeExecutionEvent {
 
     // Whether this node yielded a dynamic workflow.
     bool is_dynamic = 18;
+
+    // String location uniquely identifying where the deck HTML file is
+    // NativeUrl specifies the url in the format of the configured storage provider (e.g. s3://my-bucket/randomstring/suffix.tar)
+    string deck_uri = 19;
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:101]: Should this not be per task? What if the task retries had some decks as well

File: protos/flyteidl/service/dataproxy.proto (hunk 5 above)
Comment (EngHabu) [reply protos/flyteidl/service/dataproxy.proto:59]: Errors will return grpc errors.. there are built-in ones for Access Denied

File: protos/flyteidl/event/event.proto (hunk 6 above)
Comment (EngHabu) [reply protos/flyteidl/event/event.proto:101]: are you thinking we will need to visualize these even if the task fails?

**Review by EngHabu — APPROVED** [review]

File: protos/flyteidl/event/event.proto (hunk 6 above)
Comment (kumare3) [reply protos/flyteidl/event/event.proto:101]: Atleast maintain the flexibility to do so
